File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void reverse(vector<int>&nums,int s,int e){
        while(s<e){
            swap(nums[s++],nums[e--]);
        }
    }
    void rotate(vector<int>& nums, int k) {
        // rotating k position to left is same as rotating n-k position to the right;
        int n=nums.size();
        k=k%n;
        if(n==0 or n==1){
            return;
        }
        reverse(nums,0,n-k-1);
        reverse(nums,n-k,n-1);
        reverse(nums,0,n-1);
    }
};

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def reverse(self,nums: List[int], s: int, e: int)-> None:
        while s<e:
            nums[s],nums[e]=nums[e],nums[s]
            s+=1
            e-=1
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n=len(nums);
        k=k%n;
        self.reverse(nums,n-k,n-1);
        self.reverse(nums,0,n-k-1);
        self.reverse(nums,0,n-1);